File: plugin/src/test/java/com/oldstats/api/OldStatsApiClientTest.java
package com.oldstats.api;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.google.gson.Gson;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Protocol;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okio.Buffer;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;

public class OldStatsApiClientTest {
    private OkHttpClient httpClient;
    private Call call;
    private String serverUrl = "http://localhost:4000/";
    private String apiKey = "test-key";
    private OldStatsApiClient apiClient;

    @Before
    public void setUp() {
        httpClient = mock(OkHttpClient.class);
        call = mock(Call.class);
        when(httpClient.newCall(any())).thenReturn(call);
        apiClient = new OldStatsApiClient(httpClient, new Gson(), () -> serverUrl, () -> apiKey);
    }

    private StatEvent event() {
        return event(1);
    }

    private StatEvent event(int world) {
        return new StatEvent.WorldChange(world, Collections.emptyList());
    }

    private Response response(boolean successful) {
        return response(successful, successful ? 200 : 500);
    }

    private Response response(boolean successful, int code) {
        return new Response.Builder()
            .request(new Request.Builder().url("http://localhost:4000/api/events").build())
            .protocol(Protocol.HTTP_1_1)
            .code(code)
            .message(successful ? "OK" : "Error")
            .body(ResponseBody.create(null, "{}"))
            .build();
    }

    /** Makes callToStub's enqueue() synchronously invoke onResponse, like a same-thread fake dispatcher. */
    private void stubSuccess(Call callToStub, Response response) {
        doAnswer(invocation -> {
            invocation.<Callback>getArgument(0).onResponse(callToStub, response);
            return null;
        }).when(callToStub).enqueue(any());
    }

    private void stubFailure(Call callToStub, IOException exception) {
        doAnswer(invocation -> {
            invocation.<Callback>getArgument(0).onFailure(callToStub, exception);
            return null;
        }).when(callToStub).enqueue(any());
    }

    @Test
    public void flushingAnEmptyQueueNeverTouchesTheHttpClient() {
        apiClient.flush();
        verify(httpClient, never()).newCall(any());
    }

    @Test
    public void doesNotSendAndKeepsEventsQueuedWhenNoApiKeyIsConfigured() {
        apiKey = "";
        apiClient.enqueue(event());
        apiClient.flush();
        verify(httpClient, never()).newCall(any());

        apiKey = "now-configured";
        stubSuccess(call, response(true));
        apiClient.flush();
        verify(httpClient, times(1)).newCall(any());
    }

    @Test
    public void postsToApiEventsWithABearerTokenAndTrimsATrailingSlashFromTheServerUrl() {
        stubSuccess(call, response(true));
        apiClient.enqueue(event());
        apiClient.flush();

        ArgumentCaptor<Request> captor = ArgumentCaptor.forClass(Request.class);
        verify(httpClient).newCall(captor.capture());
        Request request = captor.getValue();
        assertEquals("http://localhost:4000/api/events", request.url().toString());
        assertEquals("Bearer test-key", request.header("Authorization"));
    }

    @Test
    public void batchesEveryQueuedEventIntoASingleRequest() {
        stubSuccess(call, response(true));
        apiClient.enqueue(event(1));
        apiClient.enqueue(event(2));
        apiClient.enqueue(event(3));
        apiClient.flush();

        verify(httpClient, times(1)).newCall(any());
    }

    @Test
    public void requeuesTheBatchForRetryWhenTheServerRejectsIt() {
        stubSuccess(call, response(false, 500));
        apiClient.enqueue(event());
        apiClient.flush(); // rejected, requeued

        Call retryCall = mock(Call.class);
        when(httpClient.newCall(any())).thenReturn(retryCall);
        stubSuccess(retryCall, response(true));
        apiClient.flush(); // should resend the same requeued event

        verify(httpClient, times(2)).newCall(any());
    }

    @Test
    public void requeuesTheBatchForRetryWhenSendingThrows() {
        stubFailure(call, new IOException("connection refused"));
        apiClient.enqueue(event());
        apiClient.flush(); // swallows the exception, requeues

        Call retryCall = mock(Call.class);
        when(httpClient.newCall(any())).thenReturn(retryCall);
        stubSuccess(retryCall, response(true));
        apiClient.flush();

        verify(httpClient, times(2)).newCall(any());
    }

    @Test
    public void flushingAgainWithNothingNewlyQueuedDoesNotSendAnEmptyRequest() {
        stubSuccess(call, response(true));
        apiClient.enqueue(event());
        apiClient.flush();
        apiClient.flush(); // queue is empty now

        verify(httpClient, times(1)).newCall(any());
    }

    @Test
    public void dropsTheOldestQueuedEventOnceTheQueueExceedsItsCap() {
        stubSuccess(call, response(true));
        // MAX_QUEUE_SIZE is 5000; fill past it so the very first event gets evicted.
        for (int i = 0; i < 5001; i++) {
            apiClient.enqueue(event(i));
        }
        apiClient.flush();

        // MAX_BATCH_SIZE is 1000, so the 5000 remaining events go out as 5 batches.
        ArgumentCaptor<Request> captor = ArgumentCaptor.forClass(Request.class);
        verify(httpClient, times(5)).newCall(captor.capture());
        StringBuilder allBodies = new StringBuilder();
        for (Request r : captor.getAllValues()) {
            allBodies.append(bodyText(r)).append('\n');
        }
        assertTrue("evicted event (world 0) must not be present", !allBodies.toString().contains("\"world\":0,"));
        assertTrue("most recent event (world 5000) must be present", allBodies.toString().contains("\"world\":5000"));
    }

    @Test
    public void capsEachRequestAtMaxBatchSizeEventsEvenWhenFarMoreAreQueued() {
        stubSuccess(call, response(true));
        for (int i = 0; i < 2500; i++) {
            apiClient.enqueue(event(i));
        }
        apiClient.flush();

        ArgumentCaptor<Request> captor = ArgumentCaptor.forClass(Request.class);
        verify(httpClient, times(3)).newCall(captor.capture());
        List<Integer> batchSizes = new ArrayList<>();
        int total = 0;
        for (Request r : captor.getAllValues()) {
            int count = bodyText(r).split("\"type\"", -1).length - 1;
            batchSizes.add(count);
            total += count;
        }
        assertTrue("no batch should exceed 1000 events, got " + batchSizes, batchSizes.stream().allMatch(s -> s <= 1000));
        assertEquals(2500, total);
    }

    private String bodyText(Request request) {
        Buffer buffer = new Buffer();
        try {
            request.body().writeTo(buffer);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return buffer.readUtf8();
    }
}

File: plugin/src/test/kotlin/com/oldstats/api/OldStatsApiClientTest.kt
package com.oldstats.api

import okhttp3.Call
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.ResponseBody
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.argumentCaptor
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.times
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever

class OldStatsApiClientTest {
    private lateinit var httpClient: OkHttpClient
    private lateinit var call: Call
    private var serverUrl = "http://localhost:4000/"
    private var apiKey = "test-key"
    private lateinit var apiClient: OldStatsApiClient

    @Before
    fun setUp() {
        httpClient = mock()
        call = mock()
        whenever(httpClient.newCall(any())).thenReturn(call)
        apiClient = OldStatsApiClient(httpClient, { serverUrl }, { apiKey })
    }

    private fun event(world: Int = 1) = StatEvent.WorldChange(world = world, worldTypes = emptyList())

    private fun response(successful: Boolean, code: Int = 200): Response =
        Response.Builder()
            .request(Request.Builder().url("http://localhost:4000/api/events").build())
            .protocol(okhttp3.Protocol.HTTP_1_1)
            .code(code)
            .message(if (successful) "OK" else "Error")
            .body(ResponseBody.create(null, "{}"))
            .build()

    @Test
    fun `flushing an empty queue never touches the http client`() {
        apiClient.flush()
        verify(httpClient, never()).newCall(any())
    }

    @Test
    fun `does not send and keeps events queued when no api key is configured`() {
        apiKey = ""
        apiClient.enqueue(event())
        apiClient.flush()
        verify(httpClient, never()).newCall(any())

        apiKey = "now-configured"
        whenever(call.execute()).thenReturn(response(successful = true))
        apiClient.flush()
        verify(httpClient, times(1)).newCall(any())
    }

    @Test
    fun `posts to api-events with a bearer token and trims a trailing slash from the server url`() {
        whenever(call.execute()).thenReturn(response(successful = true))
        apiClient.enqueue(event())
        apiClient.flush()

        val captor = argumentCaptor<Request>()
        verify(httpClient).newCall(captor.capture())
        val request = captor.firstValue
        assertEquals("http://localhost:4000/api/events", request.url().toString())
        assertEquals("Bearer test-key", request.header("Authorization"))
    }

    @Test
    fun `batches every queued event into a single request`() {
        whenever(call.execute()).thenReturn(response(successful = true))
        apiClient.enqueue(event(world = 1))
        apiClient.enqueue(event(world = 2))
        apiClient.enqueue(event(world = 3))
        apiClient.flush()

        verify(httpClient, times(1)).newCall(any())
    }

    @Test
    fun `requeues the batch for retry when the server rejects it`() {
        whenever(call.execute()).thenReturn(response(successful = false, code = 500))
        apiClient.enqueue(event())
        apiClient.flush() // rejected, requeued

        val retryCall: Call = mock()
        whenever(httpClient.newCall(any())).thenReturn(retryCall)
        whenever(retryCall.execute()).thenReturn(response(successful = true))
        apiClient.flush() // should resend the same requeued event

        verify(httpClient, times(2)).newCall(any())
    }

    @Test
    fun `requeues the batch for retry when sending throws`() {
        whenever(call.execute()).thenThrow(java.io.IOException("connection refused"))
        apiClient.enqueue(event())
        apiClient.flush() // swallows the exception, requeues

        val retryCall: Call = mock()
        whenever(httpClient.newCall(any())).thenReturn(retryCall)
        whenever(retryCall.execute()).thenReturn(response(successful = true))
        apiClient.flush()

        verify(httpClient, times(2)).newCall(any())
    }

    @Test
    fun `flushing again with nothing newly queued does not send an empty request`() {
        whenever(call.execute()).thenReturn(response(successful = true))
        apiClient.enqueue(event())
        apiClient.flush()
        apiClient.flush() // queue is empty now

        verify(httpClient, times(1)).newCall(any())
    }

    @Test
    fun `drops the oldest queued event once the queue exceeds its cap`() {
        whenever(call.execute()).thenReturn(response(successful = true))
        // MAX_QUEUE_SIZE is 5000; fill past it so the very first event gets evicted.
        for (i in 0 until 5001) {
            apiClient.enqueue(event(world = i))
        }
        apiClient.flush()

        val captor = argumentCaptor<Request>()
        verify(httpClient).newCall(captor.capture())
        val body = bodyText(captor.firstValue)
        assertTrue("evicted event (world 0) must not be present", !body.contains("\"world\":0,"))
        assertTrue("most recent event (world 5000) must be present", body.contains("\"world\":5000"))
    }

    private fun bodyText(request: Request): String {
        val buffer = okio.Buffer()
        request.body()!!.writeTo(buffer)
        return buffer.readUtf8()
    }
}
